File: apps/node-local-agent/src/verifier.ts
export class Verifier {
  private forbiddenWords: Set<string>;
  private requiredPatterns: string[];
  private minLength: number;

  constructor() {
    this.forbiddenWords = new Set<string>();
    this.requiredPatterns = ['MOCK_INFERENCE_RESULT'];
    this.minLength = 10;
  }

  verify(output: string): boolean {
    if (output.length < this.minLength) {
      return false;
    }
    for (const word of this.forbiddenWords) {
      if (output.includes(word)) {
        return false;
      }
    }
    for (const pat of this.requiredPatterns) {
      if (!output.includes(pat)) {
        return false;
      }
    }
    return true;
  }

  score(output: string): number {
    let score = 1.0;
    if (output.length < this.minLength) {
      score *= 0.5;
    }
    for (const pat of this.requiredPatterns) {
      if (output.includes(pat)) {
        score *= 1.2;
      } else {
        score *= 0.3;
      }
    }
    return Math.min(score, 1.0);
  }

  explain(output: string): string {
    if (this.verify(output)) {
      return `PASS: verified (len=${output.length}, required_patterns_matches)`;
    } else {
      return `FAIL: min_length=${this.minLength}, required=${this.requiredPatterns.length}, forbidden_check`;
    }
  }
}

File: crates/app/src/local_agent/verifier.rs
//! Deterministic verifiers for output validation.
use std::collections::HashSet;

pub struct Verifier {
    forbidden_words: HashSet<String>,
    required_patterns: Vec<String>,
    min_length: usize,
}

impl Verifier {
    pub fn new() -> Self {
        Self {
            forbidden_words: HashSet::new(),
            required_patterns: vec!["MOCK_INFERENCE_RESULT".into()],
            min_length: 10,
        }
    }

    pub fn verify(&self, output: &str) -> bool {
        if output.len() < self.min_length {
            return false;
        }
        for word in &self.forbidden_words {
            if output.contains(word) {
                return false;
            }
        }
        for pat in &self.required_patterns {
            if !output.contains(pat) {
                return false;
            }
        }
        true
    }

    pub fn score(&self, output: &str) -> f32 {
        let mut score = 1.0f32;
        if output.len() < self.min_length {
            score *= 0.5;
        }
        for pat in &self.required_patterns {
            if output.contains(pat) {
                score *= 1.2;
            } else {
                score *= 0.3;
            }
        }
        score.min(1.0)
    }

    pub fn explain(&self, output: &str) -> String {
        if self.verify(output) {
            format!("PASS: verified (len={}, required_patterns_matches)", output.len())
        } else {
            format!(
                "FAIL: min_length={}, required={}, forbidden_check",
                self.min_length,
                self.required_patterns.len()
            )
        }
    }
}
